com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/ApplicationArchiveExtractor.java

```diff
@@ -0,0 +1,200 @@
+        if (callback == null) {
+            callback = ExtractStatusCallback.NONE;
+        }
+
+        try (InputStreamProducer streamProducer = new InputStreamProducer(appArchiveStream, fileName)) {
+            return extract(streamProducer, callback);
+        } catch (IOException | SLException e) {
+            callback.onError(e);
+            throw e;
+        } catch (Exception e) {
+            callback.onError(e);
+            return null;
+        }
+    }
+
+    private Path extract(InputStreamProducer streamProducer, ExtractStatusCallback callback) throws IOException, SLException {
+        InputStream stream = streamProducer.getNextInputStream();
+        if (stream == null) {
+            throw new ContentException(com.sap.cloud.lm.sl.mta.message.Messages.CANNOT_FIND_ARCHIVE_ENTRY, streamProducer.getFileName());
+        }
+
+        if (isFile(streamProducer.getFileName())) {
+            return saveToFile(streamProducer, callback);
+        }
+
+        if (streamProducer.getFileName()
+            .equals(streamProducer.getStreamEntryName())) {
+            return saveToDirectory(streamProducer, callback);
+        }
+
+        return saveToDirectoryUnordered(streamProducer, callback);
+    }
+
+    private Path saveToFile(InputStreamProducer streamProducer, ExtractStatusCallback callback) throws IOException, SLException {
+        String prefix = getPrefix(streamProducer.getStreamEntryName());
+        String suffix = getSuffix(streamProducer.getStreamEntryName());
+
+        Path filePath = File.createTempFile(prefix, suffix)
+            .toPath();
+        callback.onFileCreated(filePath);
+        LOGGER.debug(MessageFormat.format("Saving input stream to temporary file \"{0}\"...", filePath));
+        try (OutputStream outputStream = Files.newOutputStream(filePath)) {
+            copy(streamProducer.getInputStream(), outputStream, callback);
+        }
+        LOGGER.debug(MessageFormat.format("Input stream saved to temporary file \"{0}\"", filePath));
+        return filePath;
+    }
+
+    private Path saveToDirectory(InputStreamProducer streamProducer, ExtractStatusCallback callback) throws IOException, SLException {
+        Path dirPath = getTempDirectory(streamProducer.getFileName());
+        callback.onFileCreated(dirPath);
+        LOGGER.debug(MessageFormat.format("Saving input stream to temporary directory \"{0}\"...", dirPath));
+        saveEntries(streamProducer, dirPath, callback);
+        LOGGER.debug(MessageFormat.format("Input stream saved to temporary directory \"{0}\"", dirPath));
+        return dirPath;
+    }
+
+    protected Path saveToDirectoryUnordered(InputStreamProducer streamProducer, ExtractStatusCallback callback)
+        throws IOException, SLException {
+        Path dirPath = getTempDirectory(streamProducer.getFileName());
+        callback.onFileCreated(dirPath);
+        do {
+            if (isFile(streamProducer.getStreamEntryName())) {
```

Comment:
I don't understand this. Why do this check again, when a similar one is made on line 59. What is the difference between getFileName and getStreamEntryName? How is saveToDirectoryUnordered different than saveToDirectory.

Reply:
This is recursive and is called for each file from the directory when extracting, unlike the one on line 59 which checks only a single.

Reply:
I still don't understand what's the difference between saveToDirectory and saveToDirectoryUnordered. Maybe I could, by spending several minutes reading it, but the code should express the idea more clearly. There should be no "Aha!" moment when reading the code.